### ISSUE TITLE
Cannot determine that CPU is in critical-error state from a debugger

### ISSUE THREAD

**Issue by Silabs-ArjanB** [riscv/riscv-double-trap#55]
Hello,

I (again) have some questions on the interaction between Smdbltrp and Debug.

Suppose I have an M-mode only CPU with Smdbltrp implemented and Smrnmi _not_ implemented.

For some reason the CPU enters critical-error state and at system level we do not want to use the critical-error signal to reset the chip or the CPU (as the expectation is that this will just trigger the same critical-error state again and we would have a life lock situation).

Instead we want  to connect a debugger and put the CPU into debug mode using debug _haltreq_ (so _cetrig_ is not being used). There does not seem to be any architecturally visible bit that the debugger can check to know that the CPU is in critical-error state.

What I would like is something like the following:

- New dcsr.ce bit. dcsr.ce gets set when the CPU enters critical-error state and can be read by the debugger. It can be written (e.g., cleared) from the debugger as well. If it is cleared by the debugger, then the CPU leaves critical-error state upon _dret_ and execution continues (from whatever state was updated by the debugger).

In this way a debugger can be attached to a CPU in critical-error state, the problem can be analyzed, and a work-around can be performed that would not be possible by simply resetting the CPU (which is the only way out of the critical-error state in the current Smdbltrp specification).

I realize that dcsr.ce is not useful 'in the field' (without a debugger attached), but I think that not having a way to come out of the critical-error state without mandating a rest of at least the CPU is a serious shortcoming of Smdbltrp. 

So basically I think it would be great if we could change the following:

Resuming from Debug Mode following an entry from the critical error state returns the hart to the critical error state.

could be changed into something like:

Resuming from Debug Mode following an entry from the critical error state returns the hart to the critical error state if dcsr.ce = 1 and resume (non-critical-error state) execution if dcsr.ce = 0.

(I realize that this spec has already been ratified, but maybe this behavior can be fixed in the Debug spec which has not been ratified yet)

Best regards,
Arjan

**Comment by ved-rivos** [comment]
The handling of the `critical-error` signal by the platform is implementation specific. The platform may choose to have a mode that allows an alternate handling than the default handling. The default handling in the platform are platform specific and may include actions such as a) reset the hart, b) reset the platform, c) interrupt a platform control processor to determine further action. The platform may also choose to make the `critical-error` signal visible to a debugger along with other platform specific signals made available to a debugger. 

The double trap extension specifically chose not to provide a "continue from critical error" mode to avoid burdening all implementations with the associated complexity. Further, it may not be meaningful to continue execution past this point is many cases. The debug entry on `cetrig` is expected to provide enough debug information about the path leading to the critical error to help the debugger then enter debug mode ahead of the critical error being triggered - for instance using suitable breakpoints. Needing to have a reset done before applying the work around identified was not considered significant to require the additional complexity.

**Comment by Silabs-ArjanB** [comment]
Hi @ved-rivos ,

Thank you for your fast response. I understand that the handling of the critical-error _signal_ is platform specific; that's fine.

>> Further, it may not be meaningful to continue execution past this point is many cases. 

It might indeed not be meaningful in many cases, but we consider it mandatory in some cases (and competing CPUs offer this functionality). Your described approach of entering debug first on _cetrig_, then apply some earlier breakpoint and then do a _ndmreset_ would likely work indeed (but I think that it is not as simple as it could have been).

**Comment by ved-rivos** [comment]
Agree that it may not avoid a reset in some of the cases where it was meaningful to continue execution and may need some additional steps on part of the debugger. The double trap extension does not introduce the cases that lead to double trap. Such cases existed already but would have lead to silent errors and likely been debugged. The double trap extension helps with the debug by not making them silent. 

As you noted, this extension has been ratified. So, enhancements would want to be pursued as a new extension if they provide additional value.

**Comment by Silabs-ArjanB** [comment]
Hi @ved-rivos , thanks for your feedback. You gave me some good ideas to move forward. Best regards, Arjan
